// hapi-sequelize-crud2/build/associations/one_to_many.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _createDecoratedObject(descriptors) { var target = {}; for (var i = 0; i < descriptors.length; i++) { var descriptor = descriptors[i]; var decorators = descriptor.decorators; var key = descriptor.key; delete descriptor.key; delete descriptor.decorators; descriptor.enumerable = true; descriptor.configurable = true; if ('value' in descriptor || descriptor.initializer) descriptor.writable = true; if (decorators) { for (var f = 0; f < decorators.length; f++) { var decorator = decorators[f]; if (typeof decorator === 'function') { descriptor = decorator(target, key, descriptor) || descriptor; } else { throw new TypeError('The decorator for method ' + descriptor.key + ' is of the invalid type ' + typeof decorator); } } } if (descriptor.initializer) { descriptor.value = descriptor.initializer.call(target); } Object.defineProperty(target, key, descriptor); } return target; }

var _boom = require('boom');

var _boom2 = _interopRequireDefault(_boom);

var _controller_manager = require('../controller_manager');

var _controller_manager2 = _interopRequireDefault(_controller_manager);

var _hoek = require('hoek');

var _hoek2 = _interopRequireDefault(_hoek);

var _joi = require('joi');

var _joi2 = _interopRequireDefault(_joi);

var _error = require('../error');

var _error2 = _interopRequireDefault(_error);

var _helpers = require('../helpers');

var prefix = undefined,
    scopePrefix = undefined;

var methods = {};

exports['default'] = function (server, model, association, options) {
  prefix = options.prefix;
  scopePrefix = options.scopePrefix;

  var asscOptions = _controller_manager2['default'].pluckAssociationOptions(options.controllerOptions, association._plural);

  for (var method in methods) {
    var methodOpts = asscOptions[method];

    if (!!methodOpts) {
      methodOpts = typeof methodOpts === 'object' ? methodOpts : {};

      methods[method](server, model, association, methodOpts);
    }
  }
};

var index = methods.index = function (server, model, association, options) {
  var route = _hoek2['default'].applyToDefaults(_createDecoratedObject([{
    key: 'method',
    initializer: function initializer() {
      return 'GET';
    }
  }, {
    key: 'path',
    initializer: function initializer() {
      return prefix + '/' + model._plural + '/{id}/' + association._plural;
    }
  }, {
    key: 'handler',
    decorators: [_error2['default']],
    value: function handler(request, reply) {
      var instance, _queryParams, where, offset, limit, include, result;

      return regeneratorRuntime.async(function handler$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return regeneratorRuntime.awrap((0, _helpers.getModel)(request, model));

          case 2:
            instance = context$2$0.sent;

            if (instance) {
              context$2$0.next = 5;
              break;
            }

            return context$2$0.abrupt('return', reply(_boom2['default'].notFound()));

          case 5:
            _queryParams = (0, _helpers.queryParams)(server, request);
            where = _queryParams.where;
            offset = _queryParams.offset;
            limit = _queryParams.limit;
            include = _queryParams.include;
            context$2$0.next = 12;
            return regeneratorRuntime.awrap(instance[association.accessors.get]({
              where: where,
              include: include,
              offset: offset,
              limit: limit
            }));

          case 12:
            result = context$2$0.sent;

            reply(result);

          case 14:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'config',
    initializer: function initializer() {
      return {
        validate: {
          params: {
            id: _joi2['default'].number().integer()
          },
          query: {
            filter: _helpers.validation.filter(association.target),
            include: _helpers.validation.include(association.target),
            offset: _helpers.validation.offset,
            limit: _helpers.validation.limit
          }
        }
      };
    }
  }]), options);

  server.route(route);
};

exports.index = index;
var create = methods.create = function (server, model, association, options) {
  var route = _hoek2['default'].applyToDefaults(_createDecoratedObject([{
    key: 'method',
    initializer: function initializer() {
      return 'POST';
    }
  }, {
    key: 'path',
    initializer: function initializer() {
      return prefix + '/' + model._plural + '/{id}/' + association._plural;
    }
  }, {
    key: 'handler',
    decorators: [_error2['default']],
    value: function handler(request, reply) {
      var instance, result;
      return regeneratorRuntime.async(function handler$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return regeneratorRuntime.awrap((0, _helpers.getModel)(request, model));

          case 2:
            instance = context$2$0.sent;

            if (instance) {
              context$2$0.next = 5;
              break;
            }

            return context$2$0.abrupt('return', reply(_boom2['default'].notFound()));

          case 5:
            context$2$0.next = 7;
            return regeneratorRuntime.awrap(instance[association.accessors.create](request.payload));

          case 7:
            result = context$2$0.sent;

            reply(result);

          case 9:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'config',
    initializer: function initializer() {
      return {
        validate: {
          params: {
            id: _helpers.validation.id
          }
        }
      };
    }
  }]), options);

  server.route(route);
};

exports.create = create;
var update = methods.update = function (server, model, association, options) {
  var route = _hoek2['default'].applyToDefaults(_createDecoratedObject([{
    key: 'method',
    initializer: function initializer() {
      return 'PUT';
    }
  }, {
    key: 'path',
    initializer: function initializer() {
      return prefix + '/' + model._plural + '/{id}/' + association._plural + '/{aid}';
    }
  }, {
    key: 'handler',
    decorators: [_error2['default']],
    value: function handler(request, reply) {
      var instance, result;
      return regeneratorRuntime.async(function handler$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return regeneratorRuntime.awrap((0, _helpers.getModel)(request, model));

          case 2:
            instance = context$2$0.sent;

            if (instance) {
              context$2$0.next = 5;
              break;
            }

            return context$2$0.abrupt('return', reply(_boom2['default'].notFound()));

          case 5:
            context$2$0.next = 7;
            return regeneratorRuntime.awrap(instance[association.accessors.add](request.params.aid));

          case 7:
            result = context$2$0.sent;

            reply(result);

          case 9:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'config',
    initializer: function initializer() {
      return {
        validate: {
          params: {
            id: _helpers.validation.id,
            aid: _helpers.validation.id
          }
        }
      };
    }
  }]), options);

  server.route(route);
};

exports.update = update;
var updateMany = methods.updateMany = function (server, model, association, options) {
  var route = _hoek2['default'].applyToDefaults(_createDecoratedObject([{
    key: 'method',
    initializer: function initializer() {
      return 'PUT';
    }
  }, {
    key: 'path',
    initializer: function initializer() {
      return prefix + '/' + model._plural + '/{id}/' + association._plural;
    }
  }, {
    key: 'handler',
    decorators: [_error2['default']],
    value: function handler(request, reply) {
      var instance, result;
      return regeneratorRuntime.async(function handler$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return regeneratorRuntime.awrap((0, _helpers.getModel)(request, model));

          case 2:
            instance = context$2$0.sent;

            if (instance) {
              context$2$0.next = 5;
              break;
            }

            return context$2$0.abrupt('return', reply(_boom2['default'].notFound()));

          case 5:
            context$2$0.next = 7;
            return regeneratorRuntime.awrap(instance[association.accessors.add](request.query.id));

          case 7:
            result = context$2$0.sent;

            reply(result);

          case 9:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'config',
    initializer: function initializer() {
      return {
        validate: {
          params: {
            id: _helpers.validation.id
          },
          query: {
            id: _joi2['default'].array().items(_helpers.validation.id).required()
          }
        }
      };
    }
  }]), options);

  server.route(route);
};

exports.updateMany = updateMany;
var destroy = methods.destroy = function (server, model, association, options) {
  var route = _hoek2['default'].applyToDefaults(_createDecoratedObject([{
    key: 'method',
    initializer: function initializer() {
      return 'DELETE';
    }
  }, {
    key: 'path',
    initializer: function initializer() {
      return prefix + '/' + model._plural + '/{id}/' + association._plural + '/{aid}';
    }
  }, {
    key: 'handler',
    decorators: [_error2['default']],
    value: function handler(request, reply) {
      var instance, result;
      return regeneratorRuntime.async(function handler$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return regeneratorRuntime.awrap((0, _helpers.getModel)(request, model));

          case 2:
            instance = context$2$0.sent;
            context$2$0.next = 5;
            return regeneratorRuntime.awrap(instance[association.accessors.remove](request.params.aid));

          case 5:
            result = context$2$0.sent;

            reply(result);

          case 7:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'config',
    initializer: function initializer() {
      return {
        validate: {
          params: {
            id: _helpers.validation.id,
            aid: _helpers.validation.id
          }
        }
      };
    }
  }]), options);

  server.route(route);
};

exports.destroy = destroy;
var destroyMany = methods.destroyMany = function (server, model, association, options) {
  var route = _hoek2['default'].applyToDefaults(_createDecoratedObject([{
    key: 'method',
    initializer: function initializer() {
      return 'DELETE';
    }
  }, {
    key: 'path',
    initializer: function initializer() {
      return prefix + '/' + model._plural + '/{id}/' + association._plural;
    }
  }, {
    key: 'handler',
    decorators: [_error2['default']],
    value: function handler(request, reply) {
      var instance, ids, destroyMethod, result;
      return regeneratorRuntime.async(function handler$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return regeneratorRuntime.awrap((0, _helpers.getModel)(request, model));

          case 2:
            instance = context$2$0.sent;

            if (instance) {
              context$2$0.next = 5;
              break;
            }

            return context$2$0.abrupt('return', reply(_boom2['default'].notFound()));

          case 5:
            ids = request.query.id || null;
            destroyMethod = ids ? 'removeMultiple' : 'set';
            context$2$0.next = 9;
            return regeneratorRuntime.awrap(instance[association.accessors[destroyMethod]](ids));

          case 9:
            result = context$2$0.sent;

            reply(result);

          case 11:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'config',
    initializer: function initializer() {
      return {
        validate: {
          params: {
            id: _helpers.validation.id
          },
          query: {
            id: _joi2['default'].array().items(_helpers.validation.id).optional()
          }
        }
      };
    }
  }]), options);

  server.route(route);
};

exports.destroyMany = destroyMany;
var count = methods.count = function (server, model, association, options) {
  var route = _hoek2['default'].applyToDefaults(_createDecoratedObject([{
    key: 'method',
    initializer: function initializer() {
      return 'GET';
    }
  }, {
    key: 'path',
    initializer: function initializer() {
      return prefix + '/' + model._plural + '/{id}/' + association._plural + '/count';
    }
  }, {
    key: 'handler',
    decorators: [_error2['default']],
    value: function handler(request, reply) {
      var instance, _queryParams2, where, count;

      return regeneratorRuntime.async(function handler$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return regeneratorRuntime.awrap((0, _helpers.getModel)(request, model));

          case 2:
            instance = context$2$0.sent;

            if (instance) {
              context$2$0.next = 5;
              break;
            }

            return context$2$0.abrupt('return', reply(_boom2['default'].notFound()));

          case 5:
            _queryParams2 = (0, _helpers.queryParams)(server, request);
            where = _queryParams2.where;
            context$2$0.next = 9;
            return regeneratorRuntime.awrap(instance[association.accessors.count]({ where: where }));

          case 9:
            count = context$2$0.sent;

            reply({ count: count });

          case 11:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'config',
    initializer: function initializer() {
      return {
        validate: {
          params: {
            id: _helpers.validation.id
          },
          query: {
            filter: _helpers.validation.filter(model)
          }
        }
      };
    }
  }]), options);

  server.route(route);
};
exports.count = count;
